src/test/base58_prefix_decoder_tests.cpp

```diff
@@ -0,0 +1,57 @@
+#include <base58.h>
+#include <util/base58_address.h>
+#include <boost/test/unit_test.hpp>
+
+#include <string>
+
+BOOST_AUTO_TEST_SUITE(base58_prefix_decode_test)
+
+BOOST_AUTO_TEST_CASE(base58_prefix_decode_testvectors_valid)
+{
+    const std::string base58_address_std = "12higDjoCCNXSA95xZMWUdPvXNmkAduhWv";
+
+    static const unsigned char CASES[] = {
+        0x00, // Pay to public key hash
+        0x05, // Script hash
+        0x6f, // Testnet pubkey hash
+        0xc4, // Testnet script hash
+        0x01, // Larget range of prefixes
+        0x02, // Second largest range of prefixes
+        0xff // End of possible range of inputs
+    };
+
+    static const std::vector<char> RANGES[] = {
+        {'1'},
+        {'3'},
+        {'m','n'},
+        {'2'},
+        {'Q','R','S','T','U','V','W','X','Y','Z','a','b','c','d','e','f','g','h','i','j','k','m','n','o'},
+        {'o','p','q','r','s','t','u','v','w','x','y','z','2'},
+        {'2'}
+    };
+    static_assert(std::size(CASES) == std::size(RANGES), "Base58 CASES and RANGES should have the same length");
+
+    /* common base58 address types */
+    int i = 0;
+    for (const unsigned version_byte : CASES) {
+        std::vector<unsigned char> base58_data;
+        bool valid_base58 = DecodeBase58(base58_address_std, base58_data, 35);
+        BOOST_CHECK(valid_base58  == true);
+        std::vector<char> prefixes = Base58PrefixesFromVersionByte(base58_data.size(), version_byte);
+        std::vector<char> range = RANGES[i];
+        BOOST_CHECK(prefixes == range);
+        ++i;
+    }
+
+    /* test longest segment */
+    for (unsigned short upper_range = 0x91; upper_range <= 0xFF; upper_range++) {
+        std::vector<unsigned char> base58_data;
+        bool valid_base58 = DecodeBase58(base58_address_std, base58_data, 35);
```

Comment:
This test is different since it tests the largest range of same outputs.

You can view the expected outputs at https://en.bitcoin.it/wiki/List_of_address_prefixes, The second table gives the expected results. The number 145-255 should always have a base 58 prefix of ['2'] for a 25 byte encoded length.

Reply:
An additional bonus I just pushed is 100% coverage for the space of 25 byte Base58 strings. ed4d387f5cecefe38c5e92da670a45f19426dd07